src/julia_gc.cc

```diff
@@ -210,204 +126,471 @@ static inline void * align_ptr(void * p)
-                treap = &R(*treap);
+            node->item = item;
+            return 0;
+        }
+        Int size = lsize + rsize + 1;
+        // lsize <= alpha * size && rsize <= alpha * size
+        if (alpha_hi * lsize <= alpha_lo * size &&
+            alpha_hi * rsize <= alpha_lo * size) {
+            // try further up if not unbalanced
+            return size;
+        }
+        // rebalance node
+        rebalance(node, size);
+        return 0;
+    }
+    void remove_node(Node *& node)
+    {
+        Node * del = node;
+        if (node->left != NULL) {
+            if (node->right != NULL) {
+                // copy from & delete in order successor
+                Node ** succ = &node->right;
+                while ((*succ)->left != NULL)
+                    succ = &(*succ)->left;
+                node->item = (*succ)->item;
+                remove_node(*succ);
+                return;
             }
             else {
-                treap_rot_left(treap);
-                treap = &L(*treap);
+                node = node->left;
             }
         }
-    }
-}
-
-static int treap_delete(treap_t ** treap, void * addr)
-{
-    while (*treap != NULL) {
-        int c = cmp_ptr(addr, (*treap)->addr);
-        if (c == 0) {
-            treap_delete_node(treap);
-            return 1;
+        else {
+            node = node->right;
         }
-        else if (c < 0) {
-            treap = &L(*treap);
+        dealloc(del);
+        _nodes--;
+        if (alpha_hi * _nodes <= alpha_lo * _maxnodes) {
+            if (_root)
+                rebalance(_root, _nodes);
+            _maxnodes = _nodes;
         }
+    }
+    void remove_aux(Node *& node, T & item)
+    {
+        if (!node)
+            return;
+        int c = Cmp::cmp(item, node->item);
+        if (c < 0)
+            remove_aux(node->left, item);
+        else if (c > 0)
+            remove_aux(node->right, item);
         else {
-            treap = &R(*treap);
+            remove_node(node);
         }
     }
-    return 0;
-}
 
-static uint64_t xorshift_rng_state = 1;
+  public:
+    BalancedTree()
+    {
+        init();
+    }
+    static BalancedTree * make()
+    {
+        BalancedTree<T, Cmp> * result = alloc<BalancedTree<T, Cmp> >();
+        result->init();
+        return result;
+    }
+    static void destroy(BalancedTree<T, Cmp> * tree)
+    {
+        tree->deinit();
+        dealloc(tree);
+    }
+    ~BalancedTree()
+    {
+        remove_all();
+    }
+    void init()
+    {
+        _nodes = _maxnodes = 0;
+        _root = NULL;
+        if (!height_to_size_init) {
+            height_to_size_init = 1;
+            double w = 1.0;
+            for (int d = 0; d < MaxTreeDepth; d++) {
+                w *= (double)alpha_hi;
+                w /= (double)alpha_lo;
+                height_to_size[d] = (Int)w;
+            }
+        }
+    }
+    void deinit()
+    {
+        remove_all();
+    }
+    void remove_all()
```

Comment:
```suggestion
    void clear()
```
to match STL containers?